docs: add blog about Velero + Longhorn (#1194)
longhorn/website: content/blog/20250902-k8s-backup-solutions-and-longhorn.md

# Backup Applications with Longhorn V2 Volumes using Velero

* [Overview](#overview)
* [Setup Longhorn](#setup-longhorn)
* [Setup Velero](#setup-velero)
* [Deloy Application with Longhorn V2 volume](#deloy-application-with-longhorn-v2-volume)
  * [Deploy the application](#deploy-the-application)
  * [Generate some data](#generate-some-data)
* [Backup ](#backup)
* [Restore](#restore)
* [Restore to different cluster and different storage provider](#restore-to-different-cluster-and-different-storage-provider)
* [Conclusion](#conclusion)

## Overview

When running stateful applications on Kubernetes with Longhorn as the storage provider, a common requirement is the ability to back up the entire application stack including namespaces, ConfigMaps, Secrets, PVCs/PVs. This ensure a quick and reliable restoreation if something goes wrong, even onto a new cluster.

In this guide, we’ll use Velero, a standard, open-source Kubernetes backup solution, to protect applications whose PVCs are backed by Longhorn v2 (Data Engine v2). We’ll also enable a performance optimization using Longhorn v2 linked-clone so Velero’s temporary backup PVC is created almost instantly, reducing backup time.

> Longhorn v2 linked-clone feature is introduced in Longhorn v1.10 version. Please see the Github ticket for more details: https://github.com/longhorn/longhorn/issues/7794. In a nutshell, when you clone a new PVC from a PVC using linked-clone mode, there is no data copy and the clone completes instantly.

Reviewer: ```suggestion
The Longhorn v2 linked-clone feature is introduced in the Longhorn v1.10 version. Please see the GitHub ticket for more details: https://github.com/longhorn/longhorn/issues/7794. In a nutshell, when you clone a new PVC from a PVC using linked-clone mode, there is no data copy, and the clone completes instantly.
```

> When running stateful applications on Kubernetes with Longhorn as the storage provider, a common requirement is the ability to back up the entire application stack, including namespaces, ConfigMaps, Secrets, PVCs/PVs. This ensures a quick and reliable restoration if something goes wrong, even onto a new cluster.
> 
> In this guide, we’ll use Velero, a standard, open-source Kubernetes backup solution, to protect applications whose PVCs are backed by Longhorn v2 (Data Engine v2). We’ll also enable a performance optimization using Longhorn v2 linked-clone so Velero’s temporary backup PVC is created almost instantly, reducing backup time.
> 
> The Longhorn v2 linked-clone feature is introduced in the Longhorn v1.10 version. Please see the GitHub ticket for more details: https://github.com/longhorn/longhorn/issues/7794. In a nutshell, when you clone a new PVC from a PVC using linked-clone mode, there is no data copy, and the clone completes instantly.

Reviewer 2: ```suggestion
When you run stateful applications on Kubernetes with Longhorn as the storage provider, you often need to back up the entire application stack, including namespaces, ConfigMaps, Secrets, and PVCs/PVs. This ensures a quick and reliable restoration if something goes wrong, even on a new cluster.

This guide uses Velero, a standard, open-source Kubernetes backup solution, to protect applications whose PVCs are backed by Longhorn v2 (Data Engine v2). It also enables a performance optimization using Longhorn v2 linked-clone, which creates Velero's temporary backup PVC almost instantly, reducing backup time.

The Longhorn v2 linked-clone feature was introduced in the Longhorn v1.10. For more details, see the GitHub ticket [#7794](https://github.com/longhorn/longhorn/issues/7794). In short, when you clone a new PVC from an existing PVC using linked-clone mode, no data is copied, and the clone completes instantly.
```

      namespace: velero
    data:
      node-agent-config.json: |
        {
          "backupPVC": {
            "longhorn-v2-data-engine": {
              "storageClass": "velero-backuppvc-longhorn-v2-data-engine"
            }
          }
        }
    ```

1. Install Velero using the CLI. Include the AWS object-store plugin for S3, enable CSI, and turn on the Node Agent (privileged for block-volume access). Also pass our Node Agent ConfigMap:
    ```bash
    velero install \
    --provider aws \
    --plugins velero/velero-plugin-for-aws:v1.12.2 \ # choose a compatible version with the CLI version. Ref https://github.com/vmware-tanzu/velero-plugin-for-aws?tab=readme-ov-file#compatibility
    --bucket <BUCKET> \
    --backup-location-config region=<REGION> \
    --secret-file ./credentials-velero \
    --features=EnableCSI \                        # enable CSI integration
    --use-node-agent \                            # required for data movement
    --privileged-node-agent \                     # needed for block volumes
    --node-agent-configmap=velero-node-agent-cfg  # needed for Longhorn linked-clone (fast clone)
    ```

## Deploy Application with Longhorn V2 volume

> In this section, we will deploy a fully functional [Gitea](https://about.gitea.com/) instance which is a lightweight, open-source, self-hosted Git service. We'll use a PVC backed by a Longhorn v2 volume to persist repository data. After deployment, we'll create a sample repository, and then back up the entire application stack with Velero. We'll use Longhorn v2's fast clone (linked-clone) feature to significantly speed up the backup process.
> 
> Stack components:
> 1. A dedicated namespace: `gitea-demo`
> 1. A ConfigMap and a Secret for application configuration and credentials
> 1. A Deployment for the Gitea application
> 1. A PVC backed by a Longhorn v2 volume
> 1. A Service to expose the application

Reviewer 2: ```suggestion
This section describes how to deploy a fully functional [Gitea](https://about.gitea.com/) instance, which is a lightweight, open-source, self-hosted Git service. The instance uses a PVC backed by a Longhorn v2 volume to persist repository data. After deployment, you will create a sample repository and then back up the entire application stack with Velero. This process uses Longhorn v2's fast-clone (linked-clone) feature to significantly speed up the backup.

The stack includes the following components:
```

                  key: email
          command: ["/bin/sh","-c"]
          args:
            - >
              set -e;
              /usr/local/bin/gitea migrate --config /data/gitea/conf/app.ini;
              /usr/local/bin/gitea admin user create
              --admin --username "$USERNAME" --password "$PASSWORD" --email "$EMAIL"
              --config /data/gitea/conf/app.ini || echo "admin may already exist";
          volumeMounts:
            - name: data
              mountPath: /data
      containers:
        - name: gitea
          image: docker.gitea.com/gitea:1.24.5-rootless
          ports:
            - name: http
              containerPort: 3000
          env:
            # Tell Gitea where the persistent data is
            - name: GITEA_WORK_DIR
              value: /data
            - name: GITEA_APP_INI
              value: /data/gitea/conf/app.ini
          volumeMounts:
            - name: data
              mountPath: /data

---
apiVersion: v1
kind: Service
metadata:
  name: gitea
  namespace: gitea-demo
spec:
  selector:
    app: gitea
  ports:
    - name: http
      port: 3000
      targetPort: 3000
  type: ClusterIP
```

### Generate some data

> 1. Port forward the application to localhost `kubectl -n gitea-demo port-forward svc/gitea 3000:3000`
> 1. Go to `http://localhost:3000/` to access Gitea
> 1. Log in with the username `admin` and password `ChangeMe123!`
> 1. Create a hello-world git repo and write some data to its README.md file
> 1. Clone the hello-world repo `git clone http://localhost:3000/admin/hello-world.git`
> 1. Add an OS image of around 1GiB, commit, and push it

Reviewer 2: ```suggestion
1. Port forward the application to localhost: `kubectl -n gitea-demo port-forward svc/gitea 3000:3000`
1. Access Gitea by going to: `http://localhost:3000/`.
1. Log in with the username `admin` and password `ChangeMe123!`
1. Create a "hello-world" Git repo and write some data to its `README.md` file.
1. Clone the "hello-world" repo: `git clone http://localhost:3000/admin/hello-world.git`
1. Add an OS image of around 1GiB, then commit and push it.
```

1. The outcome should be like this:
   <img src="/img/blogs/20250902-k8s-backup-solutions-and-longhorn/gitea_hello_world.png" alt="Descriptive alt text">

## Backup

> We’ll back up the entire application stack and the PVC data using CSI Snapshot Data Movement.

Reviewer 2: ```suggestion
You'll back up the entire application stack and the PVC data by using CSI Snapshot Data Movement.
```

How it works under the hood:
1. Velero creates a CSI snapshot of the PVC.
1. Velero creates a temporary backup PVC from that snapshot.
1. A data-mover pod mounts the backup PVC (read-only) and uploads the data to S3.

> Because we configured the Node Agent to use a linked-clone StorageClass, step 2 is near-instant.

Reviewer 2: ```suggestion
Since you configured the Node Agent to use a linked-clone StorageClass, step 2 is nearly instant.
```

Run this command to create a backup:
```bash
velero backup create gitea-bkp \
  --include-namespaces gitea-demo \
  --snapshot-move-data \
  --wait
```

> Observe that the cloning happens very fast. The uploading to S3 might take some time.
> 
> Once successful, you should see

Reviewer 2: ```suggestion
You should observe that the cloning happens very fast. Uploading the data to S3 may take some time.

Once the process is successful, you should see the following output:
```

```bash
➜  ~  velero backup get
NAME        STATUS      ERRORS   WARNINGS   CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
gitea-bkp   Completed   0        0          2025-09-03 19:56:14 -0700 PDT   29d       default            <none>
```

## Restore
1. To simulate a disaster case, delete the Gitea namespace so that the application stack is destroyed:
   ```bash
   kubectl delete ns gitea-demo
   ```
1. Restore the entire application stack:
   ```bash
   velero restore create gitea-restore --from-backup gitea-bkp --wait
   ```

> 1. Wait for the restoration to finish
> 1. Port forward the application to localhost

Reviewer 2: ```suggestion
1. Wait for the restoration to finish.
1. Port forward the application to localhost:
```

   ```bash
   kubectl -n gitea-demo port-forward svc/gitea 3000:3000
   ```

> 1. Go to `http://localhost:3000/` to access Gitea

Reviewer 2: ```suggestion
1. Access Gitea by going to `http://localhost:3000/`
```

1. Log in with the username `admin` and password `ChangeMe123!`

> 1. Verify that you see the same `README.md` file and the ubuntu server image

Reviewer 2: ```suggestion
1. Verify that you see the same `README.md` file and the ubuntu server image.
```

## Restore to a different cluster and different storage provider

> One very useful feature of Velero is the ability to restore the PVC data into a PVC provisioned by a different storage provider. This allows you to migrate data to different clusters and different storage providers.

Reviewer 2: ```suggestion
A very useful feature of Velero is its ability to restore PVC data into a PVC provisioned by a different storage provider. This enables you to migrate data to different clusters and storage providers.
```

1. Install Velero on the target cluster and point it to the same S3 bucket.

> 1. Map your source StorageClass → target StorageClass with this ConfigMap (in the velero namespace):

Reviewer 2: ```suggestion
1. Map your source StorageClass to your target StorageClass with this ConfigMap (in the Velero namespace):
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  # any name can be used; Velero uses the labels (below)
  # to identify it rather than the name
  name: change-storage-class-config
  # must be in the velero namespace
  namespace: velero
  # the below labels should be used verbatim in your
  # ConfigMap.
  labels:
    # this value-less label identifies the ConfigMap as
    # config for a plugin (i.e. the built-in restore item action plugin)
    velero.io/plugin-config: ""
    # this label identifies the name and kind of plugin
    # that this ConfigMap is for.
    velero.io/change-storage-class: RestoreItemAction
data:
  # add 1+ key-value pairs here, where the key is the old
  # storage class name and the value is the new storage
  # class name.
  longhorn-v2-data-engine: local-path
```

In this example, PVCs that originally used `longhorn-v2-data-engine` will be restored using `local-path` on the destination cluster.

## Conclusion

> You have now built an end-to-end application-aware backup and restore workflow for a Longhorn v2-backed workload using Velero + CSI Snapshot Data Movement. This solution provides several key benefits:

Reviewer 2: ```suggestion
You have now built an end-to-end, application-aware backup and restore workflow for a Longhorn v2-backed workload by using Velero and CSI Snapshot Data Movement. This solution provides several key benefits:
```

> * Fast backups thanks to linked-clone for the temporary backup PVC.
> * Portable, deduplicated data in S3 for cross-cluster disaster recovery.
> * Flexible restores—including remapping to a different StorageClass on a different cluster.

Reviewer 2: ```suggestion
- **Fast backups**: Thanks to linked-clone for the temporary backup PVC.
- **Portable, deduplicated data**: Stored in S3 for cross-cluster disaster recovery.
- **Flexible restores**: Including remapping to a different StorageClass on a different cluster.
```
